File: app/src/main/java/workshop/akbolatss/tools/touchcounter/utils/logging/TimberLogImplementation.java
package workshop.akbolatss.tools.touchcounter.utils.logging;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import timber.log.Timber;
import workshop.akbolatss.tools.touchcounter.BuildConfig;

public class TimberLogImplementation {

    public static void init() {
        if (BuildConfig.DEBUG)
            Timber.plant(new DevelopmentDebugTree());
        else
            Timber.plant(new ProductionTree());
    }

    public static class DevelopmentDebugTree extends Timber.DebugTree {
        @Override
        protected @Nullable String createStackElementTag(@NotNull StackTraceElement element) {
            return String.format("TouchCounter:C:%s:%s",
                    super.createStackElementTag(element),
                    element.getLineNumber());
        }
    }

    public static class ProductionTree extends Timber.Tree {

        @Override
        protected void log(int priority, @Nullable String tag, @NotNull String message, @Nullable Throwable t) {
//            if (priority == Log.ERROR || priority == Log.WARN)
//                FabricImplementation.sendStackTrace(t);
        }
    }
}

File: app/src/main/java/workshop/akbolatss/tools/touchcounter/utils/logging/TimberLogImplementation.kt
package workshop.akbolatss.tools.touchcounter.utils.logging

import timber.log.Timber
import workshop.akbolatss.tools.touchcounter.BuildConfig

object TimberLogImplementation {

    fun init() {
        val tree =
            if (BuildConfig.DEBUG) DevelopmentDebugTree()
            else ProductionTree()
        Timber.plant(tree)
    }

    private class DevelopmentDebugTree : Timber.DebugTree() {
        override fun createStackElementTag(element: StackTraceElement): String? {
            return String.format(
                "TouchCounter:C:%s:%s",
                super.createStackElementTag(element),
                element.lineNumber
            )
        }
    }

    private class ProductionTree : Timber.Tree() {
        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
//            if (priority == Log.ERROR || priority == Log.WARN)
//                Firebase.sendStackTrace(t);
        }
    }
}
